File: front-end/flutter/livee/lib/domain/repositories/campaign_repository.dart
import 'dart:convert';

import 'package:livee/data/core/api_client.dart';
import 'package:livee/domain/models/campaign.dart';
import 'package:livee/domain/models/paginated_response.dart';

class CampaignRepository {
  final ApiClient _apiClient = ApiClient();

  Future<List<Campaign>> getMyCampaigns() async {
    final response = await _apiClient.get('/campaigns/mine');
    if (response.statusCode == 200) {
      final json = jsonDecode(utf8.decode(response.bodyBytes));
      return (json['items'] as List).map((e) => Campaign.fromJson(e)).toList();
    } else {
      throw Exception('Failed to load my campaigns');
    }
  }

  Future<PaginatedResponse<Campaign>> getAllCampaigns({
    String? type,
    int? limit,
    int? page,
    String? search,
    String? sort,
  }) async {
    final Map<String, dynamic> queryParams = {};
    if (type != null) queryParams['type'] = type;
    if (limit != null) queryParams['limit'] = limit.toString();
    if (page != null) queryParams['page'] = page.toString();
    if (search != null && search.isNotEmpty) queryParams['search'] = search;
    if (sort != null) queryParams['sort'] = sort;

    final response = await _apiClient.get('/campaigns?${Uri(queryParameters: queryParams).query}');

    if (response.statusCode == 200) {
      final json = jsonDecode(utf8.decode(response.bodyBytes));
      return PaginatedResponse.fromJson(json, (itemJson) => Campaign.fromJson(itemJson));
    } else {
      throw Exception('Failed to load all campaigns');
    }
  }

  Future<Campaign> getCampaignById(String id) async {
    final response = await _apiClient.get('/campaigns/$id');
    if (response.statusCode == 200) {
      final json = jsonDecode(utf8.decode(response.bodyBytes));
      return Campaign.fromJson(json['data'] ?? json);
    } else {
      throw Exception('Failed to load campaign by id');
    }
  }

  Future<void> createCampaign(Map<String, dynamic> data) async {
    final response = await _apiClient.post('/campaigns', body: data);
    // 200번대 응답이 아닐 경우, 서버 응답 전문을 Exception으로 전달
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw Exception(utf8.decode(response.bodyBytes));
    }
  }

  Future<void> updateCampaign(String id, Map<String, dynamic> data) async {
    final response = await _apiClient.put('/campaigns/$id', body: data);
    // 200번대 응답이 아닐 경우, 서버 응답 전문을 Exception으로 전달
    if (response.statusCode != 200) {
      throw Exception(utf8.decode(response.bodyBytes));
    }
  }

  Future<void> deleteCampaign(String id) async {
    final response = await _apiClient.delete('/campaigns/$id');
    if (response.statusCode != 200) {
      throw Exception('Failed to delete campaign');
    }
  }
}

File: front-end/react/lib/domain/repositories/campaignRepository.ts
import { apiClient } from '../../apiClient';
import { Campaign } from '../models/campaign';

export class CampaignRepository {
    public async createCampaign(data: any): Promise<void> {
        try {
            await apiClient.post('/campaigns', data);
        } catch (e: any) {
            throw new Error(e.response?.data?.message || 'Failed to create campaign');
        }
    }

    public async updateCampaign(id: string, data: any): Promise<void> {
        try {
            await apiClient.put(`/campaigns/${id}`, data);
        } catch (e: any) {
            throw new Error(e.response?.data?.message || 'Failed to update campaign');
        }
    }

    public async getAllCampaigns(type?: string, limit?: number): Promise<Campaign[]> {
        const params = new URLSearchParams();
        if (type) params.append('type', type);
        if (limit) params.append('limit', limit.toString());

        try {
            const response = await apiClient.get<any>(`/campaigns?${params.toString()}`);
            // API 응답 구조에 따라 데이터 파싱
            const items = response.items || response.data?.items || response.docs || response.data?.docs || [];
            return items.map((e: any) => e as Campaign);
        } catch (error) {
            console.error('Failed to load all campaigns:', error);
            throw new Error('Failed to load all campaigns');
        }
    }

    public async getCampaignById(id: string): Promise<Campaign> {
        try {
            const response = await apiClient.get<any>(`/campaigns/${id}`);
            // API 응답 구조에 따라 데이터 파싱
            const data = response.data || response;
            return data as Campaign;
        } catch (error) {
            console.error(`Failed to load campaign with ID ${id}:`, error);
            throw new Error('Failed to load campaign details');
        }
    }

    public async getMyCampaigns(): Promise<Campaign[]> {
        try {
            const response = await apiClient.get<any>('/campaigns/mine');
            const items = response.items || [];
            return items.map((e: any) => e as Campaign);
        } catch (error) {
            console.error('Failed to load my campaigns:', error);
            throw new Error('Failed to load my campaigns');
        }
    }

    public async deleteCampaign(id: string): Promise<void> {
        try {
            await apiClient.delete(`/campaigns/${id}`);
        } catch (error) {
            console.error(`Failed to delete campaign with id ${id}:`, error);
            throw new Error('Failed to delete campaign');
        }
    }
}